File: src/nanogridbot/core/ipc_handler.py
"""IPC (Inter-Process Communication) handler for container messaging."""

import asyncio
import json
from pathlib import Path
from typing import Any

from nanogridbot.channels.base import Channel
from nanogridbot.config import get_config
from nanogridbot.database import Database


class IpcHandler:
    """Handles IPC communication with container processes."""

    def __init__(self, config: "get_config", db: Database, channels: list[Channel]):
        """Initialize IPC handler.

        Args:
            config: Application configuration
            db: Database instance
            channels: List of channel instances
        """
        self.config = config
        self.db = db
        self.channels = channels
        self._running = False
        self._watchers: dict[str, asyncio.Task] = {}

    async def start(self) -> None:
        """Start the IPC handler."""
        from loguru import logger

        self._running = True
        logger.info("IPC handler started")

        # Start watching IPC directories for all groups
        await self._watch_all_groups()

    async def stop(self) -> None:
        """Stop the IPC handler."""
        from loguru import logger

        self._running = False

        # Cancel all watchers
        for watcher in self._watchers.values():
            watcher.cancel()
            try:
                await watcher
            except asyncio.CancelledError:
                pass

        self._watchers.clear()
        logger.info("IPC handler stopped")

    async def _watch_all_groups(self) -> None:
        """Start watching IPC directories for all groups."""
        from loguru import logger

        # Get all registered groups
        groups = await self.db.get_registered_groups()

        for group in groups:
            await self._watch_group(group.jid)

        logger.info(f"Watching IPC for {len(groups)} groups")

    async def _watch_group(self, jid: str) -> None:
        """Start watching IPC directory for a group.

        Args:
            jid: Group JID
        """
        if jid in self._watchers:
            return

        task = asyncio.create_task(self._watch_group_loop(jid))
        self._watchers[jid] = task

    async def _watch_group_loop(self, jid: str) -> None:
        """Watch loop for a group's IPC directory.

        Args:
            jid: Group JID
        """
        from loguru import logger

        ipc_dir = self.config.data_dir / "ipc" / jid
        input_dir = ipc_dir / "input"
        output_dir = ipc_dir / "output"

        # Create directories
        input_dir.mkdir(parents=True, exist_ok=True)
        output_dir.mkdir(parents=True, exist_ok=True)

        # Track processed files
        processed: set[str] = set()

        while self._running:
            try:
                # Check for new input files
                for file_path in input_dir.glob("*.json"):
                    if file_path.name not in processed:
                        await self._process_input_file(jid, file_path)
                        processed.add(file_path.name)

                # Check for new output files
                for file_path in output_dir.glob("*.json"):
                    if file_path.name not in processed:
                        await self._process_output_file(jid, file_path)
                        processed.add(file_path.name)

                await asyncio.sleep(0.5)

            except asyncio.CancelledError:
                break
            except Exception as e:
                logger.error(f"IPC watch error for {jid}: {e}")
                await asyncio.sleep(1)

    async def _process_input_file(self, jid: str, file_path: Path) -> None:
        """Process an input IPC file.

        Args:
            jid: Group JID
            file_path: Path to input file
        """
        from loguru import logger

        try:
            content = file_path.read_text()
            data = json.loads(content)

            logger.debug(f"IPC input for {jid}: {data}")

            # TODO: Send to container via stdin or forward to active container

        except Exception as e:
            logger.error(f"Error processing IPC input {file_path}: {e}")

    async def _process_output_file(self, jid: str, file_path: Path) -> None:
        """Process an output IPC file.

        Args:
            jid: Group JID
            file_path: Path to output file
        """
        from loguru import logger

        try:
            content = file_path.read_text()
            data = json.loads(content)

            logger.debug(f"IPC output for {jid}: {data}")

            # Send result to channel
            result = data.get("result") or data.get("text")
            if result:
                await self._send_to_channel(jid, result)

        except Exception as e:
            logger.error(f"Error processing IPC output {file_path}: {e}")

    async def _send_to_channel(self, jid: str, text: str) -> None:
        """Send text to the appropriate channel.

        Args:
            jid: Target JID
            text: Text to send
        """
        # Find channel that owns this JID
        for channel in self.channels:
            if channel.owns_jid(jid):
                try:
                    await channel.send_message(jid, text)
                except Exception as e:
                    from loguru import logger

                    logger.error(f"Error sending to channel {channel.name}: {e}")
                break

    async def write_input(self, jid: str, sender: str, text: str) -> str:
        """Write input to IPC directory.

        Args:
            jid: Group JID
            sender: Sender name
            text: Message text

        Returns:
            Filename
        """
        from datetime import datetime

        ipc_dir = self.config.data_dir / "ipc" / jid / "input"
        ipc_dir.mkdir(parents=True, exist_ok=True)

        timestamp = datetime.now().isoformat()
        filename = f"{timestamp}.json"
        file_path = ipc_dir / filename

        file_path.write_text(
            json.dumps(
                {
                    "sender": sender,
                    "text": text,
                    "timestamp": timestamp,
                }
            )
        )

        return filename

    async def write_output(self, jid: str, result: str, session_id: str | None = None) -> str:
        """Write output to IPC directory.

        Args:
            jid: Group JID
            result: Result text
            session_id: Optional session ID

        Returns:
            Filename
        """
        from datetime import datetime

        ipc_dir = self.config.data_dir / "ipc" / jid / "output"
        ipc_dir.mkdir(parents=True, exist_ok=True)

        timestamp = datetime.now().isoformat()
        filename = f"{timestamp}.json"
        file_path = ipc_dir / filename

        data = {
            "result": result,
            "timestamp": timestamp,
        }

        if session_id:
            data["sessionId"] = session_id

        file_path.write_text(json.dumps(data))

        return filename

File: crates/ngb-core/src/ipc_handler.rs
use std::collections::HashMap;
use std::path::PathBuf;
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Arc;

use ngb_config::Config;
use ngb_types::{NanoGridBotError, Result};
use tokio::task::JoinHandle;
use tracing::{debug, error, info, warn};

/// Trait for channel adapters that can send outbound messages.
///
/// Defined here in the IPC handler because it is the primary consumer:
/// the IPC watcher routes container output to channels via this trait.
pub trait ChannelSender: Send + Sync {
    /// Return true if this sender handles the given JID.
    fn owns_jid(&self, jid: &str) -> bool;

    /// Send a text message to the specified JID.
    fn send_message(
        &self,
        jid: &str,
        text: &str,
    ) -> std::pin::Pin<Box<dyn std::future::Future<Output = Result<()>> + Send + '_>>;
}

/// File-based IPC handler that bridges containers and channel adapters.
///
/// For each watched group JID, a background task polls
/// `{data_dir}/ipc/{jid}/output/*.json` and routes results to channels.
pub struct IpcHandler {
    running: Arc<AtomicBool>,
    watchers: HashMap<String, JoinHandle<()>>,
    channels: Arc<Vec<Box<dyn ChannelSender>>>,
    data_dir: PathBuf,
    poll_interval_ms: u64,
}

impl IpcHandler {
    /// Create a new IPC handler.
    pub fn new(channels: Arc<Vec<Box<dyn ChannelSender>>>, config: &Config) -> Self {
        Self {
            running: Arc::new(AtomicBool::new(false)),
            watchers: HashMap::new(),
            channels,
            data_dir: config.data_dir.clone(),
            poll_interval_ms: 500,
        }
    }

    /// Start watching IPC output directories for the given JIDs.
    pub fn start(&mut self, jids: &[String]) {
        self.running.store(true, Ordering::SeqCst);
        info!(count = jids.len(), "Starting IPC watchers");

        for jid in jids {
            if self.watchers.contains_key(jid) {
                continue;
            }
            let output_dir = self.data_dir.join("ipc").join(jid).join("output");
            let running = self.running.clone();
            let channels = self.channels.clone();
            let jid_clone = jid.clone();
            let interval = self.poll_interval_ms;

            let handle = tokio::spawn(async move {
                watcher_loop(&jid_clone, &output_dir, &running, &channels, interval).await;
            });

            self.watchers.insert(jid.clone(), handle);
        }
    }

    /// Stop all IPC watchers.
    pub fn stop(&mut self) {
        self.running.store(false, Ordering::SeqCst);
        for (jid, handle) in self.watchers.drain() {
            handle.abort();
            debug!(jid, "Stopped IPC watcher");
        }
        info!("All IPC watchers stopped");
    }

    /// Check if the handler is running.
    pub fn is_running(&self) -> bool {
        self.running.load(Ordering::SeqCst)
    }

    /// Write an input file for a container to consume.
    ///
    /// Uses atomic write (tmp file + rename) to prevent partial reads.
    pub async fn write_input(&self, jid: &str, data: &serde_json::Value) -> Result<()> {
        let input_dir = self.data_dir.join("ipc").join(jid).join("input");
        tokio::fs::create_dir_all(&input_dir).await.map_err(|e| {
            NanoGridBotError::Container(format!("Failed to create IPC input dir: {e}"))
        })?;

        let ts = chrono::Utc::now().timestamp_millis();
        let filename = format!("input-{ts}.json");
        let tmp_path = input_dir.join(format!(".tmp-{filename}"));
        let final_path = input_dir.join(&filename);

        tokio::fs::write(&tmp_path, serde_json::to_vec_pretty(data)?)
            .await
            .map_err(|e| NanoGridBotError::Container(format!("Failed to write IPC input: {e}")))?;
        tokio::fs::rename(&tmp_path, &final_path)
            .await
            .map_err(|e| NanoGridBotError::Container(format!("Failed to rename IPC input: {e}")))?;

        debug!(jid, filename, "Wrote IPC input file");
        Ok(())
    }

    /// Write an output file (typically called by containers or tests).
    pub async fn write_output(&self, jid: &str, data: &serde_json::Value) -> Result<()> {
        let output_dir = self.data_dir.join("ipc").join(jid).join("output");
        tokio::fs::create_dir_all(&output_dir).await.map_err(|e| {
            NanoGridBotError::Container(format!("Failed to create IPC output dir: {e}"))
        })?;

        let ts = chrono::Utc::now().timestamp_millis();
        let filename = format!("output-{ts}.json");
        let tmp_path = output_dir.join(format!(".tmp-{filename}"));
        let final_path = output_dir.join(&filename);

        tokio::fs::write(&tmp_path, serde_json::to_vec_pretty(data)?)
            .await
            .map_err(|e| NanoGridBotError::Container(format!("Failed to write IPC output: {e}")))?;
        tokio::fs::rename(&tmp_path, &final_path)
            .await
            .map_err(|e| {
                NanoGridBotError::Container(format!("Failed to rename IPC output: {e}"))
            })?;

        debug!(jid, filename, "Wrote IPC output file");
        Ok(())
    }

    /// Number of active watchers.
    pub fn watcher_count(&self) -> usize {
        self.watchers.len()
    }
}

/// Background loop that polls a JID's output directory and routes messages.
async fn watcher_loop(
    jid: &str,
    output_dir: &PathBuf,
    running: &AtomicBool,
    channels: &[Box<dyn ChannelSender>],
    interval_ms: u64,
) {
    debug!(jid, dir = %output_dir.display(), "IPC watcher started");

    while running.load(Ordering::SeqCst) {
        tokio::time::sleep(std::time::Duration::from_millis(interval_ms)).await;

        let mut entries = match tokio::fs::read_dir(output_dir).await {
            Ok(e) => e,
            Err(_) => continue,
        };

        let mut files = Vec::new();
        while let Ok(Some(entry)) = entries.next_entry().await {
            let path = entry.path();
            if path.extension().is_some_and(|ext| ext == "json")
                && !path
                    .file_name()
                    .is_some_and(|n| n.to_string_lossy().starts_with('.'))
            {
                files.push(path);
            }
        }

        files.sort();

        for path in files {
            match tokio::fs::read_to_string(&path).await {
                Ok(content) => {
                    let text = extract_text(&content);
                    if !text.is_empty() {
                        route_to_channels(jid, &text, channels).await;
                    }
                    let _ = tokio::fs::remove_file(&path).await;
                }
                Err(e) => {
                    warn!(jid, path = %path.display(), error = %e, "Failed to read IPC output");
                }
            }
        }
    }

    debug!(jid, "IPC watcher stopped");
}

/// Extract text from a JSON output file.
fn extract_text(content: &str) -> String {
    if let Ok(val) = serde_json::from_str::<serde_json::Value>(content) {
        // Try common field names
        for field in ["text", "result", "message", "response"] {
            if let Some(t) = val.get(field).and_then(|v| v.as_str()) {
                return t.to_string();
            }
        }
    }
    // Fallback: return raw content if not empty JSON
    content.trim().to_string()
}

/// Send text to the first channel that owns the JID.
async fn route_to_channels(jid: &str, text: &str, channels: &[Box<dyn ChannelSender>]) {
    for channel in channels {
        if channel.owns_jid(jid) {
            if let Err(e) = channel.send_message(jid, text).await {
                error!(jid, error = %e, "Failed to send via channel");
            }
            return;
        }
    }
    warn!(jid, "No channel found for JID");
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;
    use std::sync::atomic::AtomicU32;

    /// Mock channel sender for testing.
    struct MockChannel {
        prefix: String,
        send_count: Arc<AtomicU32>,
        last_message: Arc<tokio::sync::Mutex<String>>,
    }

    impl MockChannel {
        fn new(prefix: &str) -> Self {
            Self {
                prefix: prefix.to_string(),
                send_count: Arc::new(AtomicU32::new(0)),
                last_message: Arc::new(tokio::sync::Mutex::new(String::new())),
            }
        }
    }

    impl ChannelSender for MockChannel {
        fn owns_jid(&self, jid: &str) -> bool {
            jid.starts_with(&self.prefix)
        }

        fn send_message(
            &self,
            _jid: &str,
            text: &str,
        ) -> std::pin::Pin<Box<dyn std::future::Future<Output = Result<()>> + Send + '_>> {
            let text = text.to_string();
            Box::pin(async move {
                self.send_count.fetch_add(1, Ordering::SeqCst);
                *self.last_message.lock().await = text;
                Ok(())
            })
        }
    }

    fn test_config(base: &std::path::Path) -> Config {
        Config {
            project_name: "test".to_string(),
            version: "0.0.1".to_string(),
            debug: false,
            base_dir: base.to_path_buf(),
            data_dir: base.join("data"),
            store_dir: base.join("store"),
            groups_dir: base.join("groups"),
            workspaces_dir: base.join("workspaces"),
            db_path: base.join("store/messages.db"),
            whatsapp_session_path: base.join("store/whatsapp_session"),
            openai_api_key: None,
            anthropic_api_key: None,
            telegram_bot_token: None,
            slack_bot_token: None,
            slack_signing_secret: None,
            discord_bot_token: None,
            qq_host: "127.0.0.1".to_string(),
            qq_port: 20000,
            feishu_app_id: None,
            feishu_app_secret: None,
            wecom_corp_id: None,
            wecom_agent_id: None,
            wecom_secret: None,
            dingtalk_app_key: None,
            dingtalk_app_secret: None,
            claude_api_url: "https://api.anthropic.com".to_string(),
            claude_api_version: "2023-06-01".to_string(),
            claude_model: "claude-sonnet-4-20250514".to_string(),
            claude_max_tokens: 4096,
            cli_default_group: "cli".to_string(),
            container_timeout: 300,
            container_max_output_size: 100_000,
            container_max_concurrent: 5,
            container_image: "nanogridbot-agent:latest".to_string(),
            assistant_name: "Andy".to_string(),
            trigger_pattern: None,
            poll_interval: 2000,
            max_messages_per_minute: 10,
            message_cache_size: 1000,
            batch_size: 100,
            db_connection_pool_size: 5,
            ipc_file_buffer_size: 8192,
            log_level: "INFO".to_string(),
            log_format: "default".to_string(),
            log_rotation: "10 MB".to_string(),
            log_retention: "7 days".to_string(),
            web_host: "0.0.0.0".to_string(),
            web_port: 8080,
        }
    }

    #[test]
    fn extract_text_json_text_field() {
        let content = r#"{"text": "hello world"}"#;
        assert_eq!(extract_text(content), "hello world");
    }

    #[test]
    fn extract_text_json_result_field() {
        let content = r#"{"result": "done"}"#;
        assert_eq!(extract_text(content), "done");
    }

    #[test]
    fn extract_text_plain_string() {
        assert_eq!(extract_text("just text"), "just text");
    }

    #[tokio::test]
    async fn write_input_creates_file() {
        let tmp = tempfile::tempdir().unwrap();
        let cfg = test_config(tmp.path());
        let channels: Arc<Vec<Box<dyn ChannelSender>>> = Arc::new(vec![]);
        let handler = IpcHandler::new(channels, &cfg);

        handler
            .write_input("telegram:123", &json!({"prompt": "hi"}))
            .await
            .unwrap();

        let input_dir = tmp.path().join("data/ipc/telegram:123/input");
        let mut entries = tokio::fs::read_dir(&input_dir).await.unwrap();
        let mut count = 0;
        while let Ok(Some(_)) = entries.next_entry().await {
            count += 1;
        }
        assert_eq!(count, 1);
    }

    #[tokio::test]
    async fn write_output_creates_file() {
        let tmp = tempfile::tempdir().unwrap();
        let cfg = test_config(tmp.path());
        let channels: Arc<Vec<Box<dyn ChannelSender>>> = Arc::new(vec![]);
        let handler = IpcHandler::new(channels, &cfg);

        handler
            .write_output("slack:C1", &json!({"text": "response"}))
            .await
            .unwrap();

        let output_dir = tmp.path().join("data/ipc/slack:C1/output");
        let mut entries = tokio::fs::read_dir(&output_dir).await.unwrap();
        let mut count = 0;
        while let Ok(Some(_)) = entries.next_entry().await {
            count += 1;
        }
        assert_eq!(count, 1);
    }

    #[tokio::test]
    async fn start_and_stop_watchers() {
        let tmp = tempfile::tempdir().unwrap();
        let cfg = test_config(tmp.path());
        let channels: Arc<Vec<Box<dyn ChannelSender>>> = Arc::new(vec![]);
        let mut handler = IpcHandler::new(channels, &cfg);

        assert!(!handler.is_running());
        assert_eq!(handler.watcher_count(), 0);

        handler.start(&["telegram:1".to_string(), "slack:2".to_string()]);
        assert!(handler.is_running());
        assert_eq!(handler.watcher_count(), 2);

        handler.stop();
        assert!(!handler.is_running());
        assert_eq!(handler.watcher_count(), 0);
    }

    #[tokio::test]
    async fn route_to_correct_channel() {
        let tg = MockChannel::new("telegram:");
        let sl = MockChannel::new("slack:");
        let tg_count = tg.send_count.clone();
        let sl_count = sl.send_count.clone();

        let channels: Vec<Box<dyn ChannelSender>> = vec![Box::new(tg), Box::new(sl)];

        route_to_channels("telegram:123", "hello", &channels).await;
        assert_eq!(tg_count.load(Ordering::SeqCst), 1);
        assert_eq!(sl_count.load(Ordering::SeqCst), 0);

        route_to_channels("slack:C1", "world", &channels).await;
        assert_eq!(tg_count.load(Ordering::SeqCst), 1);
        assert_eq!(sl_count.load(Ordering::SeqCst), 1);
    }
}
